[aexolate] iP

aexolate opened the pull request:
# CarbonBot
> I couldn't think of a name, so I just chose a random element

CarbonBot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2324s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😎

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aexolate committed:
Rename, Greet, Exit

aexolate committed:
added echoing, bye command

aexolate committed:
Added "list" command

aexolate committed:
Add Task.java, Implemented "mark", "unmark" commands

aexolate committed:
Add Tasks Types - Todo, Events, Deadlines

aexolate committed:
Added test case

aexolate committed:
Update linux test script

aexolate committed:
Added basic validation

aexolate committed:
Added DukeException and Error Handling

aexolate committed:
Added Delete Command

aexolate committed:
Updated Test Case for Deletions

aexolate committed:
Fix bug where Exception occurs when /to used before /from

aexolate committed:
Refactor Code, Add documentation

aexolate committed:
Add saving functionality for tasks

Save the tasks in the hard disk automatically whenever the task
list changes. Load the data from the hard disk when the chatbot
starts up.

aexolate committed:
Merge branch 'branch-Level-7'

aexolate committed:
Add Dates and Times

Allows the chatbot to understand dates and times. Task dates are
modified to be a LocalDateTime type.

aexolate committed:
Merge branch 'branch-Level-8'

aexolate committed:
Update text-ui-test

aexolate committed:
Refactor code

Reduced usage of static variables and functions to allow multiple
instance of CarbonBot.

aexolate committed:
Add Ui, Storage, TaskList

* Ui: deals with interactions with the user
* Storage: deals with loading tasks from the file and saving tasks
in the file
* TaskList: contains the task list e.g., it has operations to
add/delete tasks in the list

aexolate committed:
Add Parser, Command classes

* Parser: deals with making sense of the user command
* Added command types: AddCommand, DeleteCommand, ExitCommand,
  ListCommand, MarkCommand

aexolate committed:
Refactor main class to utilize new classes

The main class now utilizes the new Ui, Storage, Parser,
TaskList, Command classes.

aexolate committed:
Merge branch 'branch-A-MoreOOP'

aexolate committed:
Organize into packages

aexolate committed:
Merge branch 'branch-A-Packages'

aexolate committed:
Merge branch 'add-gradle-support'

aexolate committed:
Setup gradle file

aexolate committed:
Add basic JUnit Tests for Parser, TaskList

* Parser: Tests: parse()
* TaskList: Tests: delete(), get()

aexolate committed:
Modify exception message

aexolate committed:
Add JUnit tests for Storage

aexolate committed:
Merge branch 'branch-A-JUnit'

aexolate committed:
Fix MarkCommand bug

The IndexOutOfBoundsException wasn't caught properly.

aexolate committed:
Update shadowJar properties

aexolate committed:
Add checkstyle config

aexolate committed:
Change tabs to spaces

aexolate committed:
Fix more checkstyle violations

aexolate committed:
Fix TaskList imports order

aexolate committed:
Add checkstyle config

aexolate committed:
Add JavaDocs for all classes and public methods

aexolate committed:
Add find command

aexolate committed:
Add find command to Parser

aexolate committed:
Add getter for description in Task

aexolate committed:
Merge branch 'branch-A-JavaDoc'

aexolate committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/carbonbot/CarbonBot.java
#	src/main/java/carbonbot/DukeException.java
#	src/main/java/carbonbot/Parser.java
#	src/main/java/carbonbot/Storage.java
#	src/main/java/carbonbot/TaskList.java
#	src/main/java/carbonbot/Ui.java
#	src/main/java/carbonbot/command/AddCommand.java
#	src/main/java/carbonbot/command/Command.java
#	src/main/java/carbonbot/command/DeleteCommand.java
#	src/main/java/carbonbot/command/ExitCommand.java
#	src/main/java/carbonbot/command/ListCommand.java
#	src/main/java/carbonbot/command/MarkCommand.java
#	src/main/java/carbonbot/task/Deadline.java
#	src/main/java/carbonbot/task/Event.java
#	src/main/java/carbonbot/task/Task.java
#	src/main/java/carbonbot/task/Todo.java

aexolate committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/carbonbot/Parser.java

aexolate committed:
Improve code quality

aexolate committed:
Add JavaFX basic GUI

aexolate committed:
Integrate CarbonBot with the UI

aexolate committed:
Improve error messages

aexolate committed:
Add missing greetings

aexolate committed:
Fix Unicode display issue

aexolate committed:
Fix 'bye' command not exiting the program

aexolate committed:
Merge branch 'branch-Level-10'

TerrenceFyx commented on src/main/java/carbonbot/CarbonBot.java:
```diff
@@ -0,0 +1,97 @@
+    private final String saveFilePath;
+    private final Ui ui;
+    private final Storage storage;
+    private boolean shouldExit = false;
```
Consisent and logical boolean naming

TerrenceFyx commented on src/main/java/carbonbot/task/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+    @Override
+    public String serialize() {
+        return String.format("D | %d | %s | %s",
+                super.isDone ? 1 : 0,
+                super.description,
+                this.by.format(DateTimeFormatter.ofPattern("d/M/yyyy HHmm")));
```
Good split up of code

TerrenceFyx commented on src/main/java/carbonbot/Parser.java:
```diff
@@ -0,0 +1,135 @@
+     * @param fullCommand The raw string the user has input
+     * @return The Command if the fullCommand string was valid
+     */
+    public static Command parse(String fullCommand) throws DukeException {
```
Method body is abit long, perhaps you could split up into different smaller method?

TerrenceFyx commented on src/main/java/carbonbot/command/AddCommand.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        // Add the task to the TaskList
+        tasks.add(this.task);
+
+        ui.showMessage("Got it. I've added this task:");
+        ui.showMessage(this.task.toString());
+        ui.showMessage("Now you have " + tasks.size() + " tasks in the list.");
+
+        // Save the TaskList to Storage
+        try {
+            storage.write(tasks.serialize());
+        } catch (IOException ex) {
+            throw new DukeException("I/O Error: Failed to write to storage. "
+                    + ex.getMessage());
+        }
+    }
```
Good layout of logical units within a block, via a line seperation

TerrenceFyx reviewed (commented):
LGTM

jellywaiyan commented on src/main/java/carbonbot/CarbonBot.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    /**
+     * Returns whether the bot is ready to be terminated.
```
I think there should be a space between the description and the params/return section :o

jellywaiyan commented on src/main/java/carbonbot/DukeException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Thrown to indicate that an exception occurred that is specific to Duke.
+ */
+public class DukeException extends Exception {
```
If your chatbot is CarbonBot, shouldnt the exceptions be named something like CarbonBotException? (But if duke is fine then its alright too!)

jellywaiyan commented on src/main/java/carbonbot/Parser.java:
```diff
@@ -0,0 +1,135 @@
+            String from = fullCommand.substring(indexOfFrom + "/from".length(), indexOfTo).trim();
+            String to = fullCommand.substring(indexOfTo + "/to".length()).trim();
+
+            if (desc.isBlank()) {
+                throw new DukeException(":( OOPS!!! The description of an event cannot be empty.");
+            }
+            if (from.isBlank()) {
+                throw new DukeException(":( OOPS!!! The 'from' of an event cannot be empty.");
+            }
+            if (to.isBlank()) {
+                throw new DukeException(":( OOPS!!! The 'to' of an event cannot be empty.");
+            }
```
Perhaps combining these into a single if statement would be better :>

jellywaiyan commented on src/main/java/carbonbot/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+                Task task;
+
+                // Split the string by its delimiter
+                String[] cols = str.split(" \\| ");
```
Not sure what cols means, perhaps could find a better name?

jellywaiyan commented on src/main/java/carbonbot/command/MarkCommand.java:
```diff
@@ -0,0 +1,56 @@
+ */
+public class MarkCommand extends Command {
+    private final int taskIdx;
+    private final boolean isMark;
```
isMarked would be a better name imo

jellywaiyan commented on src/test/java/carbonbot/TaskListTest.java:
```diff
@@ -0,0 +1,114 @@
+        }
+
+    }
+}
```
Pretty comprehensive testing for the 3 test classes, nice :>

jellywaiyan reviewed (commented):
Pretty good! Just a few naming and style fixes, and parser could be shorted to be more clean :]

aexolate committed:
Add assertions

Several assertions are added to document important assumptions
that should hold at various points in the code.

aexolate committed:
Add exception package

To minimize code duplication and standardize exception messages,
additional exception classes were added.

DukeException is also renamed to CarbonException to remove
references to Duke.

aexolate committed:
Refactor task list handling

TaskList previously violated Single Responsibility Principle since
it was handling the parsing of data from the storage. The parsing
and deserialization of the data has now been shifted to a new class
TaskDeserializer.

TaskDeserializer takes in a line of serialized Task and returns the
deserialized Task. It also performs validation to check if the
data file is corrupted.

Storage now provide methods to directly save and fetch
TaskList. TaskDeserializer is used by Storage to save a TaskList
instance into file. The commands utilizes the new methods to reduce
code duplication.

aexolate committed:
Refactor task classes

Improved code quality of task classes by avoiding magic literals.

aexolate committed:
Replace references to Duke

All references to Duke has been replaced with Carbon or CarbonBot
to avoid confusion.

aexolate committed:
Refactor Parser

Separated the parsing logic for the different commands into their
own methods.

aexolate committed:
Refactor Ui

- Ui no longer keeps the text-based UI
- Ui now provides method to write to the GUI

- Invalid index exceptions replaced with the more
specific CarbonInvalidIndexException exception

aexolate committed:
Merge pull request #1 from aexolate/branch-A-Assertions

Add assertions

aexolate committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/carbonbot/CarbonBot.java
#	src/main/java/carbonbot/command/DeleteCommand.java
#	src/main/java/carbonbot/command/ExitCommand.java

aexolate committed:
Merge pull request #2 from aexolate/branch-A-CodeQuality

Improve code quality

aexolate committed:
Set up CI

aexolate committed:
Fix StorageTest file path not working on linux

aexolate committed:
Add load command

Load command allows the application to use a different file as the
data source.

- Usage: load <filePath>

aexolate committed:
Merge branch 'branch-C-FlexibleDataSource'

aexolate committed:
Remove faulty assert statement

aexolate committed:
Add user guide

aexolate committed:
Add screenshot

aexolate committed:
Refactor code and improve documentation

- Improve documentation
- Refactored TaskDeserializer
- Refactored TaskList & FindCommand

aexolate committed:
Fix missing file error on first run

aexolate committed:
Fix failing tests

aexolate committed:
Fix load command causing exception

aexolate committed:
Fix missing argument not throwing error

aexolate committed:
Fix javafx not included with jar file

aexolate committed:
Fix gradle javafx version
